src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILInterpreter.cs

```diff
@@ -2126,5 +2116,370 @@ private void InterpretSwitch(ref ILReader reader, int jmpBase, int[] jmpDelta)
                     reader.Seek(jmpBase + jmpDelta[i]);
             }
         }
+
+        private void InterpretNewArray(int token)
+        {
+            int length = 0;
+            StackItem stackItem = PopWithValidation();
+
+            switch (stackItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    length = stackItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    length = (int)stackItem.AsNativeInt();
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+
+            Debug.Assert(length >= 0);
+
+            TypeDesc arrayType = ((TypeDesc)_methodIL.GetObject(token)).MakeArrayType();
+            Array array = RuntimeAugments.NewArray(arrayType.GetRuntimeTypeHandle(), length);
+
+            _stack.Push(StackItem.FromObjectRef(array));
+        }
+
+        private void InterpretLoadLength()
+        {
+            Array array = (Array)PopWithValidation().AsObjectRef();
+            _stack.Push(StackItem.FromInt32(array.Length));
+        }
+
+        private void InterpretStoreElement(ILOpcode opcode)
+        {
+            StackItem valueItem = PopWithValidation();
+
+            int index = 0;
+            StackItem indexItem = PopWithValidation();
+
+            switch (indexItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    index = indexItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    index = (int)indexItem.AsNativeInt();
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+            Debug.Assert(index >= 0);
+
+            Array array = (Array)PopWithValidation().AsObjectRef();
+
+            switch (opcode)
+            {
+                case ILOpcode.stelem_i:
+                    array.SetValue(valueItem.AsNativeInt(), index);
+                    break;
+                case ILOpcode.stelem_i1:
+                    array.SetValue((sbyte)valueItem.AsInt32(), index);
+                    break;
+                case ILOpcode.stelem_i2:
+                    array.SetValue((short)valueItem.AsInt32(), index);
```

Comment:
I would've preferred to use the generic type `Array<T>` to prevent unnecessary boxing but when I do, ILC fails to compile the CoreRT console app with the following error:

```shell
Assertion Failed
  Asking for Array<T> EEType

     at ILCompiler.DependencyAnalysis.EETypeNode.OnMarked(NodeFactory context) in C:\Users\toni\Workspace\corert\src\ILCompiler.Compiler\src\Compiler\DependencyAnalysis\EETypeNode.c
s:line 1087
     at ILCompiler.DependencyAnalysisFramework.DependencyNodeCore`1.CallOnMarked(DependencyContextType context) in C:\Users\toni\Workspace\corert\src\ILCompiler.DependencyAnalysisFr
amework\src\DependencyNodeCore.cs:line 136
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.AddToMarkStack(DependencyNodeCore`1 node, String reason, DependencyNodeCore`1 reason1, DependencyNodeCore`1 reaso
n2) in C:\Users\toni\Workspace\corert\src\ILCompiler.DependencyAnalysisFramework\src\DependencyAnalyzer.cs:line 313
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1 node) in C:\Users\toni\Workspace\corert\src\ILCompiler.DependencyA
nalysisFramework\src\DependencyAnalyzer.cs:line 150
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() in C:\Users\toni\Workspace\corert\src\ILCompiler.DependencyAnalysisFramework\src\DependencyA
nalyzer.cs:line 266
     at ILCompiler.RyuJitCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in C:\Users\toni\Workspace\corert\src\ILCompiler.RyuJit\src\Compiler\RyuJitCompilation.c
s:line 40
     at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in C:\Users\toni\Workspace\corert\src\ILCompiler.Compiler\src\Compiler\Compila
tion.cs:line 387
     at ILCompiler.Program.Run(String[] args) in C:\Users\toni\Workspace\corert\src\ILCompiler\src\Program.cs:line 575
     at ILCompiler.Program.Main(String[] args) in C:\Users\toni\Workspace\corert\src\ILCompiler\src\Program.cs:line 731
```

The runtime builds successfully

Reply:
`Array<T>` is just a type that we have in the corelib so that we can write methods that support the generic interfaces on arrays. We take it's vtable and use it as the vtable of `T[]`. The type itself never exists. Even if we were to remove the check in the compiler, you wouldn't be able to cast to it at runtime because `Array<T>` is different from `T[]`.

You can call `GetRawArrayData()` to get a byref to the first element of the array and then use `Unsafe` class to do pointer arithmetic and writes/reads. Note that it will require adding bounds checks.

Reply:
`GetArrayData` is internal to `System.Private.CoreLib` so I can't use it

Reply:
I wouldn't mind making it public. The other option is to expose it through the RuntimeAugments class which is how we do it for other corelib internal APIs (reasons for RuntimeAugments are historic but we still do it for consistency).

Reply:
Is this what you have in mind?

```csharp
ref byte start = ref RuntimeAugments.GetRawDataForArray(array);
ref byte position = ref Unsafe.Add(ref start, index);

// Read from int[]
Unsafe.Read<int>(ref position);

// Write to int[]
Unsafe.Write<int>(ref position, 23);
```

Reply:
> Is this what you have in mind?

We need to multiply index by element size (available as `Array.ElementSize` in the CoreLib), but yes, that's the idea.

Reply:
Maybe use `Unsafe.As` to cast to the right element type and avoid the `Unsafe.Read` and `Unsafe.Write`. I don't know whether `Unsafe.Write` does the right write barriers if the element is a reference type.

Reply:
> I don't know whether `Unsafe.Write` does the right write barriers if the element is a reference type.

It works fine.

Reply:
Thanks! `Array.ElementSize` is definitely what I was missing